src/algebraic_geometry/projective_spectrum/structure_sheaf.lean

```diff
@@ -0,0 +1,218 @@
+
+## Main definitions and results
+* `projective_spectrum.Top`: the topological space of `projective_spectrum 𝒜` endowed with the
+  zariski topology
+* `algebraic_geometry.projective_spectrum.structure_sheaf.homogeneous_localization`: given `x` in
+  `projective_spectrum.Top 𝒜`, homogeneous localization at `x` is the subring of `Aₓ` (`A` localized
+  at prime `x`) where the numerator and denominator have same grading.
+
+Then we define the structure sheaf as the subsheaf of all dependent function
+`f : Π x : U, homogeneous_localization x` such that `f` is locally expressible as ratio of two
+elements of the *same grading*, i.e. `∀ y ∈ U, ∃ (V ⊆ U) (i : ℕ) (a b ∈ 𝒜 i), ∀ z ∈ V, f z = a / b`.
+
+* `algebraic_geometry.projective_spectrum.structure_sheaf.is_locally_fraction`: the predicate that
+  a dependent function is locally expressible as ration of two elements of the same grading.
+* `algebraic_geometry.projective_spectrum.structure_sheaf.sections_subring`: the dependent functions
+  satisfying the above local property forms a subring of all dependent functions
+  `Π x : U, homogeneous_localization x`.
+* `algebraic_geometry.Proj.structure_sheaf`: the sheaf with `U ↦ sections_subring U` and natural
+  restriction map.
+
+Then we establish that `Proj 𝒜` is a `LocallyRingedSpace`:
+* `algebraic_geometry.homogeneous_localization.is_local`: for any `x : projective_spectrum 𝒜`,
+  `homogeneous_localization x` is a local ring.
+* `algebraic_geometry.Proj.stalk_iso'`: for any `x : projective_spectrum 𝒜`, the stalk of
+  `Proj.structure_sheaf` at `x` is isomorphic to `homogeneous_localization x`.
+* `algebraic_geometry.Proj.to_LocallyRingedSpace`: `Proj` as a locally ringed space.
+
+## References
+
+* [Robin Hartshorne, *Algebraic Geometry*][Har77]
+
+
+-/
+
+noncomputable theory
+
+namespace algebraic_geometry
+
+open_locale direct_sum big_operators pointwise
+open direct_sum set_like
+
+variables {R A: Type*}
+variables [comm_ring R] [comm_ring A] [algebra R A]
+variables (𝒜 : ℕ → submodule R A) [graded_algebra 𝒜]
+
+local notation `at ` x := localization.at_prime x.as_homogeneous_ideal.to_ideal
+
+open Top topological_space category_theory opposite
+
+/--
+The underlying topology of `Proj` is the projective spectrum of graded ring `A`.
+-/
+def projective_spectrum.Top : Top := Top.of (projective_spectrum 𝒜)
+
+namespace projective_spectrum.structure_sheaf
+
+namespace homogeneous_localization
+
+open set_like.graded_monoid submodule
+
+variables {𝒜} {x : projective_spectrum.Top 𝒜}
+
+/--
+If `x` is a point in `Proj 𝒜`, then `y ∈ Aₓ` is said to satisfy `num_denom_same_deg` if and only if
+`y = a / b` where `a` and `b` are both in `𝒜 i` for some `i`.
+-/
+@[nolint has_inhabited_instance]
+structure num_denom_same_deg (y : at x) :=
+(num denom : A)
+(denom_not_mem : denom ∉ x.as_homogeneous_ideal)
+(deg : ℕ)
+(num_mem : num ∈ 𝒜 deg)
+(denom_mem : denom ∈ 𝒜 deg)
+(eq : (localization.mk num ⟨denom, denom_not_mem⟩ : at x) = y)
```

Comment:
What do you think of this alternate design?

* Remove this `eq` field and the `y` parameter
* Define `homogeneous_localization` as the quotient of `num_denom_same_deg x` by the kernel of:
  ```
  λ p, (localization.mk p.num ⟨p.denom, p.denom_not_mem⟩ : at x)
  ```

This might give you some convenient defeqs down the line, where you can use `quotient.lift` instead of `classical.some` and then having to prove the `some` satisfies the properties you care about.

Reply:
Interesting! I will experiment with this tomorrow and get back to you!

Reply:
Just to make sure that I understand you clearly, the approach is to first prove that `num_denom_same_deg x` is a ring; then use the `has_quotient` API or something similar to build `homogeneous_localization`

Reply:
> the approach is to first prove that `num_denom_same_deg x` is a ring

I'm pretty sure it isn't a ring, and can only be made a ring after taking the quotient. I'd be happy to be proved wrong though.

Reply:
I couldn't make it into a ring. 😞 
The best I can is an `add_comm_monoid`

Reply:
I couldn't make it into a ring. 😞 
The best I can is an `add_comm_monoid`

Reply:
Yes, I'm not surprised. It's like how `free_algebra.pre` is not a ring until you take the quotient.

Reply:
how about something like this
```lean
instance : add_comm_monoid (num_denom_same_deg x) := sorry

def embedding : num_denom_same_deg x →+ at x :=
{ to_fun := λ p, localization.mk p.num ⟨p.denom, p.denom_not_mem⟩,
  map_zero' := sorry,
  map_add' := sorry }

def homogeneous_localization := set.range (embedding x)
```

Reply:
I couldn't find a way of taking quotient of an add monoid by kernel. The closest thing is add group quotient by a kernel

Reply:
I'm thinking about just using `setoid.ker`, and then building the ring structure manually on the result.

Reply:
```lean
-- only bundle this if it makes the proof easier in `comm_ring` below
def embedding : num_denom_same_deg x →+ at x :=
{ to_fun := λ p, localization.mk p.num ⟨p.denom, p.denom_not_mem⟩,
  map_zero' := sorry,
  map_add' := sorry }

def homogeneous_localization := quotient (setoid.ker embedding)

instance : comm_ring (homogeneous_localization x) := sorry
```

Reply:
Elaborating a bit more on that:
```lean
def num_denom_same_deg.embedding (p : num_denom_same_deg x) : at x :=
localization.mk p.num ⟨p.denom, p.denom_not_mem⟩

def homogeneous_localization := quotient (setoid.ker $ num_denom_same_deg.embedding x)

def homogeneous_localization.val (y : homogeneous_localization x) : at x :=
quotient.lift_on' y (num_denom_same_deg.embedding x) $ λ _ _, id

def homogeneous_localization.val_surjective :
  function.injective (homogeneous_localization.val x) :=
λ a b, quotient.rec_on_subsingleton₂' a b $ λ a b h, quotient.sound' h

instance : comm_ring (homogeneous_localization x) :=
(homogeneous_localization.val_surjective x).comm_ring _ _ _ _ _ _ _
```